Mathlib/Combinatorics/SimpleGraph/Copy.lean

```diff
@@ -356,6 +360,205 @@ alias ⟨IsIndContained.exists_iso_subgraph, IsIndContained.of_exists_iso_subgra
+`SimpleGraph.copyCount` for the number of unlabelled copies. -/
+noncomputable def labelledCopyCount (G : SimpleGraph V) (H : SimpleGraph W) : ℕ := by
+  classical exact Fintype.card (Copy H G)
+
+@[simp] lemma labelledCopyCount_of_isEmpty [IsEmpty W] (G : SimpleGraph V) (H : SimpleGraph W) :
+    G.labelledCopyCount H = 1 := by
+  convert Fintype.card_unique
+  exact { default := ⟨default, isEmptyElim⟩, uniq := fun _ ↦ Subsingleton.elim _ _ }
+
+@[simp] lemma labelledCopyCount_eq_zero : G.labelledCopyCount H = 0 ↔ ¬ H ⊑ G := by
+  simp [labelledCopyCount, IsContained, Fintype.card_eq_zero_iff, isEmpty_subtype]
+
+@[simp] lemma labelledCopyCount_pos : 0 < G.labelledCopyCount H ↔ H ⊑ G := by
+  simp [labelledCopyCount, IsContained, Fintype.card_pos_iff]
+
+/-- There's more labelled copies of `H` in `G` than unlabelled ones. -/
+lemma copyCount_le_labelledCopyCount : G.copyCount H ≤ G.labelledCopyCount H := by
+  classical
+  rw [copyCount, ← Fintype.card_coe]
+  refine Fintype.card_le_of_surjective (fun f ↦ ⟨f.toSubgraph, ?_⟩) ?_
+  · simpa using ⟨f.isoToSubgraph⟩
+  · simpa [Surjective, -Copy.range_toSubgraph] using fun H' e ↦ Copy.toSubgraph_surjOn ⟨e⟩
+
+end LabelledCopyCount
+
+/-!
+#### Induced copies
+
+TODO
+
+### Killing a subgraph
+
+An important aspect of graph containment is that we can remove not too many edges from a graph `H`
+to get a graph `H'` that doesn't contain `G`.
+
+#### Killing not necessarily induced copies
+
+`SimpleGraph.killCopies G H` is a subgraph of `G` where an edge was removed from each copy of `H` in
+`G`. By construction, it doesn't contain `H` and has at most the number of copies of `H` edges less
+than `G`.
+-/
+
+private lemma aux (hH : H ≠ ⊥) {G' : G.Subgraph} :
+    Nonempty (H ≃g G'.coe) → G'.edgeSet.Nonempty := by
+  obtain ⟨e, he⟩ := edgeSet_nonempty.2 hH
+  rw [← Subgraph.image_coe_edgeSet_coe]
+  exact fun ⟨f⟩ ↦ Set.Nonempty.image _ ⟨_, f.map_mem_edgeSet_iff.2 he⟩
+
+/-- `G.killCopies H` is a subgraph of `G` where an edge was removed from each copy of `H` in
+`G`. By construction, it doesn't contain `H` (unless `H` had no edges) and has at most the number of
+copies of `H` edges less than `G`. -/
+noncomputable irreducible_def killCopies (G : SimpleGraph V) (H : SimpleGraph W) :
+    SimpleGraph V := by
+  classical exact
+  if hH : H = ⊥ then G
+  else G.deleteEdges <| ⋃ (G' : G.Subgraph) (hG' : Nonempty (H ≃g G'.coe)), {(aux hH hG').some}
+
+/-- Removing an edge from `G` for each subgraph isomorphic to `H` results in a subgraph of `G`. -/
+lemma killCopies_le_left : G.killCopies H ≤ G := by
+  rw [killCopies]; split_ifs; exacts [le_rfl, deleteEdges_le _]
+
+@[simp] lemma killCopies_bot (G : SimpleGraph V) : G.killCopies (⊥ : SimpleGraph W) = G := by
+  rw [killCopies]; exact dif_pos rfl
+
+private lemma killCopies_of_ne_bot (hH : H ≠ ⊥) (G : SimpleGraph V) :
+    G.killCopies H =
+      G.deleteEdges (⋃ (G' : G.Subgraph) (hG' : Nonempty (H ≃g G'.coe)), {(aux hH hG').some}) := by
```

Comment:
IMO it's not _ideal_ that the end-user can't access anything about the actual definition or implementation of killCopies, nor any universal property about it. But if we make it clearer in the def docstring that this is intentional, since the changes made to G are deleting completely arbitrary edges to make it happen, then it's probably fine

Reply:
I also realise now that the definition doesn't respect isos in either argument, right? That's unfortunate...

Reply:
What would be the point of making it respect isos? It sounds quite incredibly painful/impossible (because of automorphisms)

Reply:
I'm not saying you should change it to respect isos, but the majority of mathematical definitions respect isomorphisms in their inputs. I'm saying this is an exception.

Reply:
Sure, I would accept that